### PR TITLE
PLATUI-598: Store accessibility statements as YAML files

### PR DESCRIPTION
**Pull request by matthewmascord**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,13 +2,73 @@
...
+serviceKey: discounted-icecreams        # A unique name for the statement used for routing e.g. https://www.tax.service.gov.uk/accessibility-statement/discounted-icecreams
+serviceName: Discounted Icecreams       # The service name that will appear in the title of the accessibility statement
+serviceHeaderName: Icecreams            # The service name as it should appear in the grey Gov.UK header bar
+serviceDescription: |                   # A description of the service
+  Use this paragraph to describe your service. It can go over multiple lines
+  like this.
+serviceDomain: www.tax.service.gov.uk   # The domain name underwhich this service exists (exclude the https:// and the path)
+serviceUrl: /icecreams                  # The relative URL to the service (omitting www.tax.service.gov.uk)
+contactFrontendServiceId: icecreams    # The service id passed to contact-frontend and hmrc-deskpro
+complianceStatus: partial               # Whether this service is fully or partially compliant
```
Comment (JoPintoPaul) [inline README.md:19]: Maybe list the allowed values, i.e. Should be `fully|partial`

File: README.md [hunk 2]
```diff
@@ -2,13 +2,73 @@
...
+accessibilityProblems:
+  - a description of the first problem
+  - another description of a problem
+  - a description of the last problem
+milestones:
+  - description: The X page does not have Y and also does not display Z. This doesn't meet WCAG success
+      criterion X.Y.Z (Criterion Description).
+    date: 2020-10-31                    # The date that this issue will be fixed by
```
Comment (JoPintoPaul) [inline README.md:27]: Be good to be explicit with date format, i.e. `yyyy-mm-dd`

File: README.md [hunk 3]
```diff
@@ -2,13 +2,73 @@
...
+  - description: |
+      The service has errors in X and incorrectly nested Z which means assistive technologies
+      cannot use the service reliably. This does not meet WCAG success criterion
+      X.Y.Z (Criterion Description).
+    date: 2020-09-30                    # The date that this issue will be fixed by
+serviceSendsOutboundMessages: false     # Set to true if the service sends or asks for documents from service users
+serviceLastTestedDate: 2019-09-15
+statementVisibility: public             # If set to public, the statement will be visible in production
+statementCreatedDate: 2019-09-30
+statementLastUpdatedDate: 2019-09-30
```
Comment (JoPintoPaul) [inline README.md:37]: Again, suggest being explicit with date formats.

File: README.md [hunk 4]
```diff
@@ -2,13 +2,73 @@
...
+services:
+  - challenge-a-childcare-decision
+  - coronavirus-job-retention-scheme
+  - direct-debit
+  - online-payments
+  - disguised-remuneration
+  - pay-what-you-owe-in-instalments
+  - discounted-icecreams              # It's crucial this key matches the serviceKey in your service's statement YAML file
+``` 
+
```
Comment (JoPintoPaul) [inline README.md:58]: Do we want to specify if they should insert it at the beginning, at the end, or in alphabetical order?

File: README.md (hunk 1 above)
Comment (matthewmascord) [reply README.md:19]: Would it be worth including an example of a fully compliant statement?

File: app/uk/gov/hmrc/accessibilitystatementfrontend/repos/AccessibilityStatementsRepo.scala [hunk 5]
```diff
@@ -16,35 +16,33 @@
...
-      statementCreatedDate = new GregorianCalendar(2019, Calendar.SEPTEMBER, 23).getTime,
-      statementLastUpdatedDate = new GregorianCalendar(2019, Calendar.APRIL, 1).getTime
-    )
-  )
+case class AccessibilityStatementsSourceRepo @Inject()(
+  appConfig: AppConfig,
+  statementsParser: AccessibilityStatementsParser,
+  statementParser: AccessibilityStatementParser)
+    extends AccessibilityStatementsRepo {
+  import appConfig._
```
Comment (JoPintoPaul) [inline app/uk/gov/hmrc/accessibilitystatementfrontend/repos/AccessibilityStatementsRepo.scala:35]: Maybe move this import up to be with its buddies?

File: app/uk/gov/hmrc/accessibilitystatementfrontend/repos/AccessibilityStatementsRepo.scala [hunk 6]
```diff
@@ -16,35 +16,33 @@
...
+
+  private val accessibilityStatements: Map[String, AccessibilityStatement] = {
+    val services = statementsParser.parseFromSource(statementsSource).valueOr(throw _).services
+    val statements = services map { service =>
+      statementParser.parseFromSource(statementSource(service)).valueOr(throw _)
+    }
+
+    (services zip statements).toMap
```
Comment (JoPintoPaul) [inline app/uk/gov/hmrc/accessibilitystatementfrontend/repos/AccessibilityStatementsRepo.scala:43]: Question on the functionality of `zip` - I can't remember, does it just work off of the ordering? If so, we might want to construct the Map slightly more verbosely.

File: app/uk/gov/hmrc/accessibilitystatementfrontend/views/StatementPage.scala.html [hunk 7]
```diff
@@ -20,6 +20,7 @@
 @this(layout: Layout)
 
 @(statement: AccessibilityStatement)(implicit request: Request[_], messages: Messages, appConfig: AppConfig)
+@serviceAbsoluteUrl = {https://@statement.serviceDomain@statement.serviceUrl}
```
Comment (JoPintoPaul) [inline app/uk/gov/hmrc/accessibilitystatementfrontend/views/StatementPage.scala.html:23]: Question on this one: in other places, I've been putting values like this within the class. Might be good to agree a consistent approach?

File: project/AppDependencies.scala [hunk 8]
```diff
@@ -4,21 +4,30 @@ import sbt.Keys.libraryDependencies
...
 
 object AppDependencies {
+  private val circeVersion = "0.12.0"
+
+  private val circeDependencies = Seq(
+    "io.circe" %% "circe-core",
+    "io.circe" %% "circe-generic",
+    "io.circe" %% "circe-parser",
+    "io.circe" %% "circe-yaml"
+  ).map(_ % circeVersion)
```
Comment (JoPintoPaul) [inline project/AppDependencies.scala:14]: Honestly, if it were me, I'd just put these in with the other compile dependencies as they're not re-used by the test dependencies atm, but it's not a hill I'll die on.

File: test/unit/config/ServicesSpec.scala [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      }
+    }
+  }
+
+  "the configuration files" should {
+    services.foreach { (service: String) =>
+      s"include a correctly formatted accessibility statement yaml file for $service" in {
+        val source = appConfig.statementSource(service)
+
+        statementParser.parseFromSource(source).valueOr(throw _)
```
Comment (JoPintoPaul) [inline test/unit/config/ServicesSpec.scala:60]: Is this test just checking that no error is thrown?

File: app/uk/gov/hmrc/accessibilitystatementfrontend/repos/AccessibilityStatementsRepo.scala (hunk 5 above)
Comment (matthewmascord) [reply app/uk/gov/hmrc/accessibilitystatementfrontend/repos/AccessibilityStatementsRepo.scala:35]: This was importing the members of appConfig which isn't available at the top.

File: app/uk/gov/hmrc/accessibilitystatementfrontend/repos/AccessibilityStatementsRepo.scala (hunk 6 above)
Comment (matthewmascord) [reply app/uk/gov/hmrc/accessibilitystatementfrontend/repos/AccessibilityStatementsRepo.scala:43]: Yeah that's correct. It takes corresponding elements from each list and forms pairs: https://www.scala-lang.org/api/current/scala/collection/Seq.html#zip[B](that:scala.collection.IterableOnce[B]):CC[(A@scala.annotation.unchecked.uncheckedVariance,B)]

File: app/uk/gov/hmrc/accessibilitystatementfrontend/views/StatementPage.scala.html (hunk 7 above)
Comment (matthewmascord) [reply app/uk/gov/hmrc/accessibilitystatementfrontend/views/StatementPage.scala.html:23]: You mean the AccessibilityStatement case class? I would vote for keeping rendering concerns away from the case class an in the template, so the case class is just about providing the raw data.

File: test/unit/config/ServicesSpec.scala (hunk 9 above)
Comment (matthewmascord) [reply test/unit/config/ServicesSpec.scala:60]: Yeah, do you think I should be more explicit about that?

File: README.md (hunk 4 above)
Comment (matthewmascord) [reply README.md:58]: I've said at the end, do you have a preference?

File: README.md (hunk 1 above)
Comment (JoPintoPaul) [reply README.md:19]: I don't think there needs to be a full example in the README, but I think where there is a list of allowed values, that should definitely be in the README.

File: README.md (hunk 4 above)
Comment (JoPintoPaul) [reply README.md:58]: At the end is fine with me, but maybe make that a bit more explicit? I missed that in the first reading.

**Pull request merged**